python/pyarrow/tests/test_types.py

```diff
@@ -330,3 +343,35 @@ def test_field_equality_operators():
     assert f1 != f3
     assert f3 != f4
     assert f1 != 'foo'
+
+
+def test_field_metadata():
+    f1 = pa.field('a', pa.int8())
+    f2 = pa.field('a', pa.int8(), metadata={})
+    f3 = pa.field('a', pa.int8(), metadata={b'bizz': b'bazz'})
+
+    assert f1.metadata is None
+    assert f2.metadata == {}
+    assert f3.metadata[b'bizz'] == b'bazz'
+
+
+def test_field_add_remove_metadata():
+    f0 = pa.field('foo', pa.int32())
+
+    assert f0.metadata is None
+
+    metadata = {b'foo': b'bar', b'pandas': b'badger'}
+
+    f1 = f0.add_metadata(metadata)
```

Comment:
`add_metadata` is a bit misleading. It actually sets the metadata instead of adding new pairs to it.

Reply:
Then we should maybe deprecate this method and replace with `replace_metadata`, `set_metadata` or `with_metadata`. Knowing Spark's immutable API, I would have chosen `with_metadata` but I guess this is probably not what the average Python user would expect as naming.

Reply:
Both in tha Python and C++ api? Created issue https://issues.apache.org/jira/browse/ARROW-2769